rfd/0015-2fa-management.md

```diff
@@ -0,0 +1,269 @@
+----------------   ----   -------------------------------   -------------------------------
+solokey            U2F    Wed 16 Dec 2020 02:05:46 PM PST   Wed 16 Dec 2020 02:05:46 PM PST
+my-otp-app         OTP    Wed 16 Dec 2020 02:06:46 PM PST   Wed 16 Dec 2020 02:06:46 PM PST
+
+# If 2FA is optional:
+$ tsh mfa rm solokey
+You are about to remove the only remaining MFA device.
+This will disable MFA during login.
+Are you sure? (y/N): N
+
+# If 2FA is required:
+$ tsh mfa rm solokey
+Can't remove the only remaining MFA device.
+Please add a replacement MFA device first using "tsh mfa add".
+```
+
+#### web UI
+
+Web UI management of 2FA devices should be logically similar to the CLI:
+- a page to see all enrolled devices
+- a button to enroll a new device
+- buttons to remove any enrolled device (except for the last one)
+
+Web UI details, wireframes and implementation will be added later, when we have
+the capacity to do it. Initially, 2FA management is CLI-only.
+
+#### bootstrap
+
+Initially, a user account doesn't have a 2FA device. Depending on [cluster
+configuration](#configuration), use of 2FA devices might be required.
+
+If 2FA is required, a user is required to enroll a device during account
+creation (for local users) or first login (for SSO users).
+
+If 2FA is optional, a user can create an account and login without 2FA. They
+can then add 2FA devices as described above. If an existing user has at least
+one 2FA device registered, it's required during login.
+
+### configuration
+
+The current 2FA configuration in Teleport only applies to local users and is
+always required. We need to allow this to be optional (for users to migrate)
+and allow to mix different 2FA methods.
+
+Existing configuration options must be backwards-compatible - no change in
+behavior unless config values are changed.
+
+New values for `auth_service.authentication.second_factor` for this:
+- `off` (existing) - no 2FA can be enrolled
+- `otp` (existing) - only OTP can be enrolled and is required for all local
+  users
+- `u2f` (existing) - only U2F can be enrolled and is required for all local
+  users
+- `on` (new) - users can enroll both OTP and U2F devices, and 2FA is required
+  for all local users
+- `optional` (new) - users can enroll both OTP and U2F devices, and 2FA is
+  required only for users with 2FA enrolled
+
+#### restricted device vendors
+
+Another new option is restrictions on U2F device manufacturers. This is done
+using attestation certificates presented by the device during enrollment. See
+[FIDO
+docs](https://fidoalliance.org/specs/fido-u2f-v1.2-ps-20170411/fido-u2f-overview-v1.2-ps-20170411.html#verifying-that-a-u2f-device-is-genuine)
+about attestation.
+
+On the Teleport side, users can pass the trusted attestation CAs as so:
+
+```yaml
+# teleport.yaml
+auth_service:
+  authentication:
+    second_factor: "on" # or "u2f" or "optional"
+    u2f_device_attestation_cas:
```

Comment:
@awly are we going to support those as resources via API as well as config parameters? what's the data model?

Reply:
This is stored in the `AuthPreference` of the auth server: https://github.com/gravitational/teleport/blob/4b5bb54bf4390bdcc7c8b45424a248998c42d7e4/api/types/authentication.go#L267-L290 (along with the other U2F parameters).
The actual "storage" is the `teleport.yaml` of the auth server. There are REST endpoints for `AuthPreference` https://github.com/gravitational/teleport/blob/4b5bb54bf4390bdcc7c8b45424a248998c42d7e4/lib/auth/apiserver.go#L201-L202. It seems to be read-only in practice (only tests call `SetAuthPreference`).

`u2f_device_attestations_cas` are not in the same boat as our host/user/jwt CAs. These U2F CAs are just the certificates (not keys), usually from 3rd parties like Yubico.

Reply:
@awly can you modify the design to be API friendly? In the current form it will be impossible to use it with cloud or terraform provider using API. I can imagine for example to support paths and content as optional parameter, so folks can set up auth preference resource using API and supply CAs

Reply:
Added an example of passing the raw CA PEM blob in the config.
It can also be passed via `SetAuthPreference` remotely.
Is this what you mean?